crypto/ake/enclave/src/lib.rs

```diff
@@ -554,14 +555,16 @@ impl<EI: EnclaveIdentity> AkeEnclaveState<EI> {
     pub fn backend_decrypt(
         &self,
         responder_id: ResponderId,
-        msg: EnclaveMessage<ClientSession>,
+        msg: EnclaveMessage<NonceSession>,
     ) -> Result<Vec<u8>> {
         // Ensure lock gets released as soon as we're done decrypting.
         let mut backends = self.backends.lock()?;
         backends
             .get_mut(&responder_id)
             .ok_or(Error::NotFound)
-            .and_then(|session| Ok(session.decrypt(&msg.aad, &msg.data)?))
+            .and_then(|session| {
+                Ok(session.decrypt_with_nonce(&msg.aad, &msg.data, msg.channel_id.peek_nonce())?)
```

Comment:
Should this be `get_nonce` or `peek_nonce`? I figured it should be peek nonce because we're just using it to decrypt and I think the nonce should only increment when it encrypts a message back to the sender?

Reply:
(I'm coming in fresh so probably am lacking some understanding)
I think peeking makes sense.  If somewhere else in the code has a need to decrypt `msg` then a get here will prevent future decryptions.

One thing I'm not clear on is the message back.  Most of the steps look to be post incrementing the nonce.  This means unless there is an explicit pre-increment the first reply might re-use the same nonce.

Reply:
My belief is that either `peek_nonce` or `get_nonce` would work here, as we're taking ownership of the `EnclaveMessage<NoiseSession>` and not returning it, and `NoiseSession` doesn't hold the nonce value by reference or anything like that.

As for Nick's comment, `decrypt_with_nonce` calls `set_nonce` and then `decrypt`, which will post-increment the nonce. So the only case in which we'd be possibly in trouble would be a sequence that looks like this:

`Node1 sends message A with nonce 0 -> Node2 decrypts message A with nonce 0 -> Node2 sends message B with nonce 1 -> Node2 decrypts message A with nonce 0 (for a second time) -> Node2 sends message C with nonce 1 (key is now compromised)`.

Is `EnclaveMessage<NonceSession>` `Clone` (or `Copy` I guess)? Because if so, the above can happen (although it would have to be us that screwed it up rather than client code, so at least that's nice).

Reply:
Updating from offline: `peek_nonce()` is fine in this case, because `channel_id` is ephemeral, and as a general rule, `const > mut`.

It's probably advisable to refactor `peek_nonce()` into an immutable `nonce()` method, and delete `get_nonce()` altogether. I wrote those accessors before I decided apps didn't need to shoulder the responsibility for maintaining nonces apart from `Ready<>`.

Reply:
SGTM.